plugins/synaprom/fu-synaprom-device.c

```diff
@@ -0,0 +1,504 @@
+ * transferring iotas with multiple commands */
+typedef struct __attribute__((packed)) {
+	guint8		 command;
+	guint16		 itype;		/* type of iotas to find */
+	guint16		 flags;		/* flags, see below */
+	guint8		 maxniotas;	/* maximum number of iotas to return, 0 = unlimited */
+	guint8		 firstidx;	/* first index of iotas to return */
+	guint8		 dummy[2];
+	guint32		 offset;	/* byte offset of data to return */
+	guint32		 nbytes;	/* maximum number of bytes to return */
+} FuSynapromCmdIotaFind;
+
+typedef struct __attribute__((packed)) {
+	guint16		 status;
+} FuSynapromReplyGeneric;
+
+typedef struct __attribute__((packed)) {
+	guint16		 status;
+	guint32		 fullsize;	/* size of all data */
+} FuSynapromReplyIotaFind;
+
+typedef struct __attribute__((packed)) {
+	guint16		 status;
+	guint32		 buildtime;	/* Unix-style build time */
+	guint32		 buildnum;	/* build number */
+	guint8		 vmajor;	/* major version */
+	guint8		 vminor;	/* minor version */
+	guint8		 target;	/* target, e.g. VCSFW_TARGET_ROM */
+	guint8		 product;	/* product, e.g.  VCSFW_PRODUCT_FALCON */
+	guint8		 siliconrev;	/* silicon revision */
+	guint8		 formalrel;	/* boolean: non-zero -> formal release */
+	guint8		 platform;	/* Platform (PCB) revision */
+	guint8		 patch;		/* patch level */
+	guint8		 serial_number[6]; /* 48-bit Serial Number */
+	guint8		 security[2];	/* bytes 0 and 1 of OTP */
+	guint32		 patchsig;	/* opaque patch signature */
+	guint8		 iface;		/* interface type, see below */
+	guint8		 otpsig[3];	/* OTP Patch Signature */
+	guint16		 otpspare1;	/* spare space */
+	guint8		 reserved;	/* reserved byte */
+	guint8		 device_type;	/* device type */
+} FuSynapromReplyGetVersion;
+
+/* the following bits describe security options in
+** FuSynapromReplyGetVersion::security[1] bit-field */
+#define FU_SYNAPROM_SECURITY1_PROD_SENSOR		(1 << 5)
+
+/* this is followed by a chain of iotas, as follows */
+typedef struct __attribute__((packed)) {
+	guint16		 status;
+	guint16		 nbytes;
+	guint16		 itype;
+} FuSynapromReplyIotaFindHdr;
+
+/* this iota contains the configuration id and version */
+typedef struct __attribute__((packed)) {
+	guint32		 config_id1;	/* YYMMDD */
+	guint32		 config_id2;	/* HHMMSS */
+	guint16		 version;
+	guint16		 unused[3];
+} FuSynapromConfigVersion;
+
+typedef struct __attribute__((packed)) {
+	guint32		 product;
+	guint32		 id;		/* MFW unique id used for compat verification */
+	guint32		 buildtime;	/* unix-style build time */
+	guint32		 buildnum;	/* build number */
+	guint8		 vmajor;	/* major version */
+	guint8		 vminor;	/* minor version */
+	guint8		 unused[6];
+} FuSynapromMfwHeader;
+
+G_DEFINE_TYPE (FuSynapromDevice, fu_synaprom_device, FU_TYPE_USB_DEVICE)
+
+static gboolean
+fu_synaprom_device_open (FuUsbDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (device);
+	if (!g_usb_device_claim_interface (usb_device, 0x0,
+					   G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
```

Comment:
Is this the same interface that fprint would need?  If so, then is this potentially going to be problematic when fprint tries to interact with it too?

Reply:
Possibly; I don't know a good way around this. I think print should probably rebind okay if the kernel driver gets detached and attached; it should just be like the device is removed and re-inserted in a few ms. I'll do some testing but I'm not overly worried.

Reply:
I guess you'll need to check the fprint implementation claims and hangs onto it the entire time the daemon is running or only on demand. 
If it's only on demand the "easiest" way around it to me is make sure that both the fprint implementation and fwupd implementation will schedule a retry a little later if the device is claimed by one or the other.

The other solution that comes to mind would be for fprint to to "share" a firmware updating interface over to fwupd.  So when fwupd runs it can call an inhibit method in fprint that lets fprint let go of the device for fwupd to probe it or update it.

Reply:
I talked to @hadess about this a bit today: tl;dr: printd only claims the device when you try to use it, for instance run an image capture. Claiming the device on insertion for a split second and also for the duration of the firmware update is going to be fine.

Reply:
> Claiming the device on insertion for a split second and also for the duration of the firmware update is going to be fine.

It's possible that the interactions might feel a bit weird if somebody gains privileges through pam_fprintd to trigger the firmware update, depending on timing (if you tried to start the update when the fingerprint interaction hasn't quite finished, or authentication is needed in the middle of an update). But there should be no chances of the two processes clobbering each other.